File: abc355b/abc355b.ml
open Printf
open Scanf

let rec merge lx ly =
  match lx, ly with
    x :: rx, y :: ry when x < y -> 0 :: merge rx ly
  | x :: rx, y :: ry -> 1 :: merge lx ry
  | rx, [] -> List.map (fun _ -> 0) rx
  | [], ry -> List.map (fun _ -> 1) ry

let solve la lb =
  let l = merge (List.sort compare la) (List.sort compare lb) in
  let rec loop l =
    match l with
      x :: y :: ry when x = 0 && y = 0 -> "Yes"
    | x :: y :: ry -> loop (y :: ry)
    | _ -> "No"
  in loop l

let () =
  scanf "%d %d " (fun n m ->
    let la = List.init n @@ fun _ -> scanf "%d " @@ fun x -> x in
    let lb = List.init m @@ fun _ -> scanf "%d " @@ fun x -> x in
    solve la lb) |> printf "%s\n"

File: abc355b/abc355b.hs
import Data.List(sort, find)

main :: IO ()
main = getLine >> solve <$> f <*> f >>= putStrLn
  where
    f = map read <$> words <$> getLine

solve :: [Int] -> [Int] -> String
solve as bs = maybe "No" (const "Yes") $ find (\(x, y) -> x == 0 && y == 0) $ zip ls (tail ls)
  where
    ls = merge (sort as) (sort bs)

merge :: [Int] -> [Int] -> [Int]
merge (a:as) (b:bs)
  | a < b = 0 : merge as (b:bs)
  | otherwise = 1 : merge (a:as) bs
merge [] bs = replicate (length bs) 1
merge as [] = replicate (length as) 0
